Implemented the first eigenvalue example

fdkong opened the pull request:
I have created the following hierarchy:

```
             +--> DisplacedSystem
             | 
SystemBase ->+--> AuxiliarySystem
             |
             +-->NonlinearSystemBase+--> NonlinearSystem
                                    |
                                    +--> NonlinearEigenSystem
```

```
SubProblem ->+--> DisplacedProblem
             |
             +-->FEProblemBase+--> FEProblem
                              |
                              +--> EigenProblem
```

And integrated with SLEPc to create the first standard eigenvalue example.

Before I further add more functionalities, and I would ask suggestions and comments from @idaholab/moose-team.  The implementation will affect the MOOSE a lot.  I want to know if it is a `MOOSE` way to write the code like this. And any improvements on this implementation?

fdkong committed:
Make most of classes use NonlinearSystemBase

Make most of classes use NonlinearSystemBase instead of NonlinearSystem. NonlinearSystemBase will
serve as a basis class for both NonlinearSystem and NonlinearEigenSystem.

Some classes do not need to use NonlinearSystem at all, and in this situation, NonlinearSystem is removed
totally from the classes. It makes code cleaner.

Refs #7398

fdkong committed:
Moved copyOldSolution and copyOlderSolution down

to NonlinearSystem. Because it depends on the libmesh systems.

It will be different for Eigensystem and NonlinearImplicitSystem.

Refs #7398

fdkong committed:
Get rid of TransientNonlinearImplicitSystem from

NonlinearSystemBase. In other words, we moved TransientNonlinearImplicitSystem
down to NonlinearSystem because different subclass of NonlinearSystemBase
has different libmesh system behind it.

Functions: solve(), setupFiniteDifferencedPreconditioner(), converged(), have been
made as pure functions, and the implementation is left for subclasses: NonlinearSystem
and NonlinearEigenSystem.

Added functions: RHS() and nonlinearSolver() into NonlinearSystemBase  to wrapper data from Libmesh system.

Normally, other classes should use system() instead of sys() because the class, System, returned from the function
system() is the superclass of all libmesh systemes.

Refs #7398

fdkong committed:
Renamed getNonlinearSystem to getNonlinearSystemBase

FEProblem will serve as a basis class for EquationProblem and
EigenProblem. Most of sys() has been replaced by system().

Refs #7398

fdkong committed:
Removed redundant code from computeResidual

and computeJacobian in FEProblem.c

Refs #7398

fdkong committed:
Created a new class  EquationProblem

Refs #7398

fdkong committed:
Changed default problem type to EquationProblem

Refs #7398

fdkong committed:
Changed _nl in FEProblem to pointer * _nl

so that we do not need to initialize during the constructor.

Later on, the child class EquationProblem creates and setup
NonlinearSystem.

Refs #7398

fdkong committed:
Removed _nl dependent code from FEProblem constructor

to EquationProblem so that we can delay the construction
of NonlinearSystem.

Refs #7398

fdkong committed:
Moved the constructions of NonlinearSystem and

AuxiliarySystem to EquationProblem. The reason is that
we will use different problem class to create different
system. For EigenProlbem should create NonlinearEigenSystem
and EquationProblem creates NonlinearSystem.

The problem type could be specified in input file, and EquationProblem
will be used by default.

Refs #7398

fdkong committed:
Created a new class NonlinearEigenSystem which

use TransientEigenSystem from libmesh

Refs #7398

fdkong committed:
Created a new class EigenProblem

which takes care of NonlinearEigenSystem

Refs #7398

fdkong committed:
The first eigenvalue example using SLEPc

Integrated with SLEPc to make the first eigenvalue
example. The example is a standard eigenvalue problem.

Refs # 7398

YaqiWang commented:
@fdkong Just to make sure in you first 11 commit, you are doing exactly as the message says but not sneaking anything in. 

How about we change the `NonlinearSystemBase` back to `NonlinearSystem` and rename `NonlinearSystem` to `NonlinearEquationSystem`?

I will look into it in more detail.

YaqiWang commented:
I can suggest to get the names by the member functions on the level. Just for example, if one level is constructing the mesh, we can name it `MeshProblem`.

fdkong committed:
Fixed issues after resolved conflicts

Refs #7398

fdkong commented:
> I can suggest to get the names by the member functions on the level. Just for example, if one level is constructing the mesh, we can name it MeshProblem.

According suggestions from @YaqiWang  and @jwpeterson, I am going to change `EquationProblem` to `FEProblem` (keep the same API), and `FEProblem` to `MeshProblem`. `SubProblem` will be merged into `Problem`.

@idaholab/moose-team, everybody agrees on this? @permcody, @friedmud, @andrsd ?

friedmud commented:
Please don't change the interface for the class called `FEProblem`.  A LOT of user code relies on it (including LOTS of my own).

To change something like that would require a lengthy deprecation period (6 months?)... with a clear migration path.

Lots of people depend on MOOSE now... many outside the apps you guys see daily.  Gone is the time when we could make major API changes and just "deal with" the fall out.

fdkong commented:
> Please don't change the interface for the class called FEProblem. A LOT of user code relies on it (including LOTS of my own).
> 
> To change something like that would require a lengthy deprecation period (6 months?)... with a clear migration path.
> 
> Lots of people depend on MOOSE now... many outside the apps you guys see daily. Gone is the time when we could make major API changes and just "deal with" the fall out.

OK, @jwpeterson has pointed out yesterday, and I will keep the same API. Do you have a suggestion on what name we have for the basis class of `FEProblem` and `EigenProblem`?

permcody commented:
I haven't really been following this thread but I know this is a lot of work and I think it's looking pretty good. Thanks for responding to everyone's comments.

andrsd commented on framework/src/base/EigenProblem.C:
```diff
@@ -0,0 +1,135 @@
+    FEProblem(parameters),
+    _nl_eigen(new NonlinearEigenSystem(*this, "eigen0"))
+{
+  _nl =  _nl_eigen;
```
1 space after `=`.

andrsd commented on framework/src/base/EigenProblem.C:
```diff
@@ -0,0 +1,135 @@
+    _nl_eigen(new NonlinearEigenSystem(*this, "eigen0"))
+{
+  _nl =  _nl_eigen;
+  _aux = new AuxiliarySystem(*this, "aux0");
```
Both `_nl` and `_aux` could be set in the initializer list...

andrsd commented on framework/src/base/EigenProblem.C:
```diff
@@ -0,0 +1,135 @@
+EigenProblem::solve()
+{
+  Moose::perf_log.push("Eigen_solve()", "Execution");
+#ifdef LIBMESH_HAVE_PETSC
```
I believe that `LIBMESH_HAVE_SLEPC` implies `LIBMESH_HAVE_PETSC`

andrsd commented on framework/src/base/EigenProblem.C:
```diff
@@ -0,0 +1,135 @@
+    _nl->solve();
+
+  if (_solve)
+    _nl->update();
```
You can merge this into one `if`

andrsd commented on framework/src/base/EigenProblem.C:
```diff
@@ -0,0 +1,135 @@
+bool
+EigenProblem::converged()
+{
+  _console<<"WARNING: did not implement yet \n";
```
Spaces around `<<`

andrsd commented on framework/src/base/EquationProblem.C:
```diff
@@ -0,0 +1,78 @@
+    FEProblem(parameters),
+    _use_nonlinear(getParam<bool>("use_nonlinear"))
+{
+  _nl = _use_nonlinear? (new NonlinearSystem(*this, "nl0")) : (new MooseEigenSystem(*this, "eigen0"));
```
Spaces around `?`

andrsd commented on framework/src/base/EquationProblem.C:
```diff
@@ -0,0 +1,78 @@
+    _use_nonlinear(getParam<bool>("use_nonlinear"))
+{
+  _nl = _use_nonlinear? (new NonlinearSystem(*this, "nl0")) : (new MooseEigenSystem(*this, "eigen0"));
+  _aux = new AuxiliarySystem(*this, "aux0");
```
Both `_nl` and `_aux` can be set in the initializer list.

andrsd commented on framework/src/base/FEProblem.C:
```diff
@@ -3618,12 +3595,11 @@ FEProblem::computeTransientImplicitJacobian(Real time, const NumericVector<Numbe
+  _nl->duDotDu() = shift;
   _time = time;
-  computeJacobian(sys,u,jacobian);
+  computeJacobian(u,jacobian);
```
Space after a `,`

andrsd commented on framework/src/base/NonlinearSystem.C:
```diff
@@ -151,61 +77,18 @@ namespace Moose {
-    _has_nodalbc_save_in(false),
-    _has_nodalbc_diag_save_in(false)
+  NonlinearSystemBase(fe_problem, fe_problem.es().add_system<TransientNonlinearImplicitSystem>(name), name),
+  _transient_sys(fe_problem.es().get_system<TransientNonlinearImplicitSystem>(name))
```
Indetation

andrsd commented on framework/src/outputs/VariableResidualNormsDebugOutput.C:
```diff
@@ -57,7 +57,7 @@ VariableResidualNormsDebugOutput::output(const ExecFlagType & /*type*/)
   for (unsigned int var_num = 0; var_num < _sys.n_vars(); var_num++)
   {
-    Real var_res_id = _sys.calculate_norm(*_sys.rhs,var_num,DISCRETE_L2);
+    Real var_res_id = _sys.calculate_norm(_problem_ptr->getNonlinearSystemBase().RHS(),var_num,DISCRETE_L2);
```
Space after `,`s

andrsd commented on framework/src/preconditioners/PhysicsBasedPreconditioner.C:
```diff
@@ -254,7 +254,7 @@ PhysicsBasedPreconditioner::apply(const NumericVector<Number> & x, NumericVector
     //Copy rhs from the big system into the small one
     MoosePreconditioner::copyVarValues(mesh,
-        _nl.sys().number(),system_var,x,
+        _nl.system().number(),system_var,x,
```
Space after `,`s. Also below

andrsd commented on framework/src/timesteppers/DT2.C:
```diff
@@ -61,29 +61,29 @@ void
+  System & nl_sys = _fe_problem.getNonlinearSystemBase().system();
+  _u1 = & nl_sys.add_vector("u1", true, GHOSTED);
+  _u2 = & nl_sys.add_vector("u2", false, GHOSTED);
+  _u_diff = & nl_sys.add_vector("u_diff", false, GHOSTED);
```
`&nl_sys`, no need for additional space, similarly below

andrsd commented on framework/src/timesteppers/DT2.C:
```diff
@@ -61,29 +61,29 @@ void
   // save solution vectors
-  *_u_saved = *nl_sys.current_local_solution;
-  *_u_older_saved = *nl_sys.older_local_solution;
+  *_u_saved = * nl_sys.currentSolution();
```
`*nl_sys`

andrsd commented on framework/src/utils/PetscSupport.C:
```diff
@@ -137,7 +137,7 @@ setSolverOptions(SolverParams & solver_params)
 
 void
-petscSetupDM (NonlinearSystem & nl) {
+petscSetupDM (NonlinearSystemBase & nl) {
```
`{` on the new line

andrsd commented on test/tests/problems/eigen_problem/ipm.i:
```diff
@@ -0,0 +1,100 @@
+ uniform_refine = 0
+
+ displacements = 'x_disp y_disp'
+[]
```
Indentation

andrsd commented on test/tests/problems/eigen_problem/ipm.i:
```diff
@@ -0,0 +1,100 @@
+#Its inverse will be 0.49950700634518.
+
+[Variables]
+  active = 'u'
```
Remove this

andrsd commented on test/tests/problems/eigen_problem/ipm.i:
```diff
@@ -0,0 +1,100 @@
+[]
+
+[Kernels]
+  active = 'diff rea'
```
Remove

andrsd commented on test/tests/problems/eigen_problem/ipm.i:
```diff
@@ -0,0 +1,100 @@
+  active = 'homogeneous'
+
+  [./homogeneous]
+    type = DirichletBC
```
Remove

andrsd commented on test/tests/problems/eigen_problem/ipm.i:
```diff
@@ -0,0 +1,100 @@
+[Problem]
+  type = EigenProblem
+  n_eigen_pairs = 5;
+  n_basis_vectors = 15;
```
`;`s?

andrsd reviewed (changes requested):
I went roughly over the code, not in gore details.
1. I am not a big fan of the private members in the new FEProblem-derived classes.
2. I did not comment on the class name change that was already discussed, because I agree with the outcome.

andrsd commented:
Also, I saw quite a bit of `not implemented yet`, what is the plan there?

fdkong commented:
> Also, I saw quite a bit of not implemented yet, what is the plan there?

There are somethings I highly possibly (not certainly ) will  touch, and they are actually pure virtual functions in the basis class, and I don't want them keep silent at this point if they are called in other places.

YaqiWang commented on framework/include/base/EigenProblem.h:
```diff
@@ -0,0 +1,51 @@
+  NonlinearEigenSystem *_nl_eigen;
+};
+
+#endif /* LIBMESH_HAVE_SLEPC */
```
Our current eigen solver does not depend on SLEPc. We have a decision to make here, should we keep this SLEPc independent solver on our side or push it to SLEPc and make SLEPc mandatory for solving eigenvalue problems?

fdkong commented:
> When people do not have SLEPc build, we should provide an "empty" EigenProblem class which will throw an error in a c-tor saying something like "You are trying to solve an eigenvalue problem, but you do not have SLEPc built in. Go recompile libmesh with such and such options."

Good suggestion, and I will do this.

fdkong commented:
> Our current eigen solver does not depend on SLEPc. We have a decision to make here, should we keep this SLEPc independent solver on our side or push it to SLEPc and make SLEPc mandatory for solving eigenvalue problems?

@YaqiWang, if you want to keep the current eigen-solver on our side, I am fine with that, but it is difficult for me to maintain the current implementation.   I prefer to rewrite the current algorithms (inverse power iteration and Newton) based on SLPEc.

YaqiWang commented:
Understood. This will make SLPEc required by Rattlesnake. I am not sure if it is a good idea for political reasons. Just want to warn you. We need others' opinion especially @permcody @friedmud. I personally do not like this but am willing to be persuaded.

fdkong commented:
> Understood. This will make SLPEc required by Rattlesnake. I am not sure if it is a good idea for political reasons. Just want to warn you. We need others' opinion especially @permcody @friedmud. I personally do not like this but am willing to be persuaded.

If for some reasons we are not willing to include SLEPc as part of MOOSE-based applications, I definitely can build some eigenvalue solvers  based on PETSc only (of course, more work for me).  Current SLEPc also has some issues on the nonlinear eigenvalue problems, that is, it does not support eigen-vector dependent nonlinear eigenvalue  problems. More precisely, operators can not depend on the eigen vector. I have to create new algorithms, even though  we bring in SLEPc. Of course, benefit here is we could use some components from SLEPc when implementing the new algorithms. Another benefit is that we have a lot more algorithms than right now if we bring in SLEPc.

Ping @friedmud, @permcody.

friedmud commented:
I don't see a reason to lose our current eigenvalue solver capability.
It's very mature and can solve very difficult problems.

Power iterations are key in many nuclear reactor applications... we're
always going to need the ability to do that manually.  For instance, I use
a power iteration based eigenvalue solver in my new method of
characteristics code.  Also: acceleration methods like CMFD require power
iterations.

Not everything we do can boil down to calling slepc.  I think we want to
_add_ slepc as a new option to see how it performs with some problems.

On Fri, Nov 4, 2016 at 12:36 PM Fande Kong <notifications@github.com> wrote:

> Understood. This will make SLPEc required by Rattlesnake. I am not sure if
> it is a good idea for political reasons. Just want to warn you. We need
> others' opinion especially @permcody <https://github.com/permcody>
> @friedmud <https://github.com/friedmud>. I personally do not like this
> but am willing to be persuaded.
>
> If for some reasons we are not willing to include SLEPc as part of
> MOOSE-based applications, I definitely can build some eigenvalue solvers
> based on PETSc only (of course, more work for me). Current SLEPc also has
> some issues on the nonlinear eigenvalue problems, that is, it does not
> support eigen-vector dependent nonlinear eigenvalue problems. More
> precisely, operators can not depend on the eigen vector. I have to create
> new algorithms, even though we bring in SLEPc. Of course, benefit here is
> we could use some components from SLEPc when implementing the new
> algorithms. Another benefit is that we have a lot more algorithms than
> right now if we bring in SLEPc.
>
> Ping @friedmud <https://github.com/friedmud>, @permcody
> <https://github.com/permcody>.
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/pull/8009#issuecomment-258481877>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AA1JMQoRjyNv0IM0bzyiKNqxXSPrdQtyks5q618DgaJpZM4Ko6rp>
> .
>

fdkong commented:
> I don't see a reason to lose our current eigenvalue solver capability. It's very mature and can solve very difficult problems. 

The reason is that it is not easy to maintain the current implementation. I am not going to remove the current capability until SLEPc option can do the same thing. 

If @YaqiWang  does NOT like make his apps depend on SLEPc, I could re-implement the two algorithms (inverse and Newton) based on PETSc only. 

If @friedmud  still want to keep the current implementation after I have re-implemented the same algorithms using PETSc or SLEPc, I am fine too because I don't consider they can't exist together.  

>Power iterations are key in many nuclear reactor applications... we're always going to need the ability to do that manually. For instance, I use a power iteration based eigenvalue solver in my new method of characteristics code. Also: acceleration methods like CMFD require power iterations. Not everything we do can boil down to calling slepc. I think we want to _add_ slepc as a new option to see how it performs with some problems.

friedmud commented:
If you can really do all of the things that the current EigenSystem does
then it should be fine.

Note: I don't actually use the current EigenSystem stuff (and I agree that
it's complicated)... I just don't want to lose sight of the fact that it
solves things that I don't think slepc can do on its own... and I would
hate to see that capability go away.

On Fri, Nov 4, 2016 at 5:45 PM Fande Kong <notifications@github.com> wrote:

> I don't see a reason to lose our current eigenvalue solver capability.
> It's very mature and can solve very difficult problems.
>
> The reason is that it is not easy to maintain the current implementation.
> I am not going to remove the current capability until SLEPc option can do
> the same thing.
>
> If @YaqiWang <https://github.com/YaqiWang> does NOT like make his apps
> depend on SLEPc, I could re-implement the two algorithms (inverse and
> Newton) based on PETSc only.
>
> If @friedmud <https://github.com/friedmud> still want to keep the current
> implementation after I have re-implemented the same algorithms using PETSc
> or SLEPc, I am fine too because I don't consider they can't exist together.
>
> Power iterations are key in many nuclear reactor applications... we're
> always going to need the ability to do that manually. For instance, I use a
> power iteration based eigenvalue solver in my new method of characteristics
> code. Also: acceleration methods like CMFD require power iterations. Not
> everything we do can boil down to calling slepc. I think we want to *add*
> slepc as a new option to see how it performs with some problems.
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/pull/8009#issuecomment-258555464>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AA1JMQNyFtkohcUXukqOtOy8zRTXgMyjks5q66dbgaJpZM4Ko6rp>
> .
>

fdkong commented:
> If you can really do all of the things that the current EigenSystem does then it should be fine. Note: I don't actually use the current EigenSystem stuff (and I agree that it's complicated)... I just don't want to lose sight of the fact that it solves things that I don't think slepc can do on its own... and I would hate to see that capability go away.

I know SLEPc can not do that, but I am going to build something on top of SLEPc and PETSc to make the capability. The capability we have right now will NOT go away, and will be enhanced, otherwise no reason to do this development.

fdkong commented:
> Both _nl and _aux could be set in the initializer list...

There are some data members from basis class, so we could not put them in the initializer list, (I am not sure).

andrsd commented:
> There are some data members from basis class, so we could not put them in the initializer list, (I am not sure).

Yeah, I saw that after I commented on it. One way how to deal with it is to move the data members into child classes and make the parent class use some virtual method to get the right instance of nl or aux... Just an idea...

fdkong committed:
Addressed format issues

Refs #7398

moosebuild commented:
Job `Test intel:linux-intel` on 9df694f : invalidated by @fdkong

moosebuild commented:
Job `Test intel:linux-intel` on b28a8ba : invalidated by @fdkong

fdkong committed:
Replace "FEProblem" with "FEProblemBase" in

SwapBackSentinel.h and ComputeIndicatorThread.C to

resolved conflicts and to fix compiling issues

Refs #7398

moosebuild commented:
Job `App tests:linux-gnu` on 3a49aa3 : invalidated by @fdkong

moosebuild commented:
Job `App tests:linux-gnu` on 3a49aa3 : invalidated by @fdkong

fdkong committed:
Deprecated  interaces MultiApp::problem()

MultiApp::appProblem(), FEProblemBase::getNonlinearSystem()
and ActionWarehouse::problem()

Refs #7398

moosebuild commented:
Job `App tests:linux-gnu` on 3a49aa3 : invalidated by @fdkong

moosebuild commented:
Job `App tests:linux-gnu` on 3a49aa3 : invalidated by @fdkong

moosebuild commented:
Job `App tests:linux-gnu` on 3a49aa3 : invalidated by @fdkong

Mammoth

moosebuild commented:
Job `App tests:linux-gnu` on 3a49aa3 : invalidated by @brianmoose

Mammoth should be fixed.

fdkong commented:
@permcody, @andrsd, finally passed all tests.

fdkong committed:
Rename NonlinearSystem.{C,h} -> NonlinearSystemBase.{C,h}

fdkong committed:
Rename NonlinearSystem -> NonlinearSystemBase in code.

fdkong committed:
Add new almost empty NonlinearSystem class.

fdkong committed:
Rename FEProblem.{C,h} to FEProblemBase.{C, h}

fdkong committed:
Rename EquationProblem.{C,h} to FEProblem.{C,h}

fdkong committed:
Change actual code (rename class names and so on) in FEProblemBase.{C,h}

fdkong committed:
Change the actual code in FEProblem.{C,h}

Rename class name EquationProblem to FEProblem

fdkong committed:
Replace "FEProblem" with "FEProblemBase" in

the whole moose

Refs #7398

fdkong committed:
Add SLEPC safeguard to NonlinearEigenSystem.{C,h}

fdkong committed:
Added getNonlinearSystem() back to FEProblemBase.C

And also added a wrapper setInputParametersFEProblem() to FEProblemBase.C
and FEProblemBase.C to make sure parameters.set<FEProblemBase *>("_fe_problem_base")
and parameters.set<FEProblemBase *>("_fe_problem") are  called
at the same time.

Refs #7398

fdkong committed:
Added appProblem() and problem() back to MultiApp.{C,h}

to make the interface consistent with users' apps.

And added the new interfaces appProblemBase() and problemBase().

Refs #7398

fdkong committed:
Added interface problem() back to ActionWarehouse.{C,h}

and also added a new interface problemBase().

Refs #7398

fdkong committed:
Change transfers to user appProblemBase() and problemBase()

Refs #7398

fdkong committed:
Replace #include "FEProblemBase.h"  with "FEProblem.h"

because users are using the class FEProblem.

Replace parameters.XXX<FEProblemBase *>("_fe_problem") with parameters.XXX<FEProblemBase *>("_fe_problem_base")

Refs #7398

permcody commented:
@fdkong - we'll review this. However, we should also go ahead and run "clang-format" on the files affected by this PR before we merge (as your last commit). You may as well keep the ownership of these lines since we'll run that command on the whole repo very soon. @rwcarlsen can probably help with that.

fdkong commented:
> @fdkong - we'll review this. However, we should also go ahead and run "clang-format" on the files affected by this PR before we merge (as your last commit). You may as well keep the ownership of these lines since we'll run that command on the whole repo very soon. @rwcarlsen can probably help with that.

We delay to this because clang-format can create annoying things

fdkong commented:
@friedmud, more suggestions on this?

andrsd commented on framework/include/base/EigenProblem.h:
```diff
@@ -0,0 +1,48 @@
+  virtual void outputStep(ExecFlagType type) override;
+
+protected:
+  NonlinearEigenSystem *_nl_eigen;
```
space after `*`

andrsd commented on framework/include/splits/Split.h:
```diff
@@ -45,7 +45,7 @@ class Split :
   };
 
-  FEProblem& _fe_problem;
+  FEProblemBase& _fe_problem;
```
Space before `&`.

andrsd commented on framework/src/actions/ActionWarehouse.C:
```diff
@@ -367,3 +368,10 @@ ActionWarehouse::printInputFile(std::ostream & out)
+ActionWarehouse::problem()
+{
+  mooseDeprecated("ActionWarehouse::problem() is deprecated, please use ActionWarehouse::problemBase() \n");
+  return std::dynamic_pointer_cast<FEProblem >(_problem);
```
No space inside `<...>`

andrsd commented on framework/src/actions/AddOutputAction.C:
```diff
@@ -58,8 +58,8 @@ AddOutputAction::act()
-  // Add a pointer to the FEProblem class
-  _moose_object_pars.addPrivateParam<FEProblem *>("_fe_problem",  _problem.get());
+  // Add a pointer to the FEProblemBase class
+  _moose_object_pars.addPrivateParam<FEProblemBase *>("_fe_problem_base",  _problem.get());
```
2 spaces in front of `_problem.get()`

andrsd commented on framework/src/base/MooseApp.C:
```diff
@@ -386,7 +386,7 @@ MooseApp::runInputFile()
   if (error_unused || warn_unused)
   {
-    MooseSharedPointer<FEProblem> fe_problem= _action_warehouse.problem();
+    MooseSharedPointer<FEProblemBase> fe_problem= _action_warehouse.problemBase();
```
Spaces around `=`

andrsd commented on framework/src/base/NonlinearEigenSystem.C:
```diff
@@ -0,0 +1,122 @@
+  _transient_sys(fe_problem.es().get_system<TransientEigenSystem>(name))
+#else
+  NonlinearSystemBase(fe_problem, fe_problem.es().add_system<TransientBaseSystem>(name), name),
+  _transient_sys(fe_problem.es().get_system<TransientBaseSystem>(name))
```
4 spaces for the initializer list...

andrsd commented on framework/src/base/NonlinearEigenSystem.C:
```diff
@@ -0,0 +1,122 @@
+{
+  #ifndef LIBMESH_HAVE_SLEPC
+    mooseError("Need to install SLEPc to solve eigenvalue problems, please reconfigure\n");
+  #endif /* LIBMESH_HAVE_SLEPC */
```
Incorrect indentation. << 2

andrsd commented on framework/src/base/NonlinearEigenSystem.C:
```diff
@@ -0,0 +1,122 @@
+
+
+ NumericVector<Number> &
+ NonlinearEigenSystem::RHS()
```
Indent by 1 space?

andrsd commented on framework/src/multiapps/MultiApp.C:
```diff
@@ -341,15 +341,34 @@ MultiApp::getBoundingBox(unsigned int app)
   unsigned int local_app = globalAppToLocal(app);
 
-  return _apps[local_app]->getExecutioner()->feProblem();
+  return dynamic_cast<FEProblem&>(_apps[local_app]->getExecutioner()->feProblem());
```
`FeProblem &`

andrsd commented on framework/src/outputs/VariableResidualNormsDebugOutput.C:
```diff
@@ -57,7 +57,7 @@ VariableResidualNormsDebugOutput::output(const ExecFlagType & /*type*/)
   for (unsigned int var_num = 0; var_num < _sys.n_vars(); var_num++)
   {
-    Real var_res_id = _sys.calculate_norm(*_sys.rhs,var_num,DISCRETE_L2);
+    Real var_res_id = _sys.calculate_norm(_problem_ptr->getNonlinearSystemBase().RHS(), var_num,DISCRETE_L2);
```
The other comma needs a space after it as well...

andrsd commented on framework/src/restart/Restartable.C:
```diff
@@ -15,13 +15,17 @@
+    _restartable_subproblem(parameters.isParamValid("_subproblem") ?
+    parameters.get<SubProblem *>("_subproblem") : (parameters.isParamValid("_fe_problem_base") ?
+    parameters.get<FEProblemBase *>("_fe_problem_base") : (parameters.isParamValid("_fe_problem") ?
+    parameters.get<FEProblem *>("_fe_problem") : subproblem)) )
```
This needs better indentation to increase legibility. I propose:
```
_var(bool1 ? long_thing :
  (bool2 ? longer_thing :
    (bool3 ? the_longer_things : what ever)))
```
But anything else that will clearly show what is going on with the nesting will work too. Also, since it is a pointer a nested `if` in the `c-tor` could look a lot better.

andrsd commented on framework/src/timesteppers/TimeStepper.C:
```diff
@@ -31,7 +31,7 @@ InputParameters validParams<TimeStepper>()
     MooseObject(parameters),
     Restartable(parameters, "TimeSteppers"),
-    _fe_problem(*parameters.getCheckedPointerParam<FEProblem *>("_fe_problem")),
+    _fe_problem(parameters.have_parameter<FEProblemBase *>("_fe_problem_base") ? *getParam<FEProblemBase *>("_fe_problem_base") : *getParam<FEProblem *>("_fe_problem")),
```
I assuming this is here for backwards compatibility?

fdkong replied:
Yes,  some applications explicitly use `parameter.xxx<FEProblem *>("_fe_problem")`

andrsd commented:
One comment on the `EigenSystem` and using preprocessor. The following typically worked better for me:

In .h:
```
#if HAVE_A
  declare class X as if we have A
#else
  declare class X as if we do not have A 
    - keep to c-tor signature the same, but does not require inheritance
    - does not even need to have the same interface!
#endif
```

In .C
```
#if HAVE_A
  implement class X as if we have A
#else
  implement class X as if we do not have A
#endif
```

This obviously depends on how we want X to behave if we don't have A. If our goal is to just throw an error and finish when people try to use X, we end up with just one c-tor. IMO, it is more elegant than to try pick and `#ifdef` individual lines or group of lines.

fdkong commented:
> One comment on the EigenSystem and using preprocessor. The following typically worked better for me:
> 
> In .h:
> 
> #if HAVE_A
>   declare class X as if we have A
> #else
>   declare class X as if we do not have A 
>     - keep to c-tor signature the same, but does not require inheritance
>     - does not even need to have the same interface!
> #endif
> 
> In .C
> 
> #if HAVE_A
>   implement class X as if we have A
> #else
>   implement class X as if we do not have A
> #endif
> 
> This obviously depends on how we want X to behave if we don't have A. If our goal is to just throw an error and finish when people try to use X, we end up with just one c-tor. IMO, it is more elegant than to try pick and #ifdef individual lines or group of lines.

I will update according to your suggestions. Thanks a lot!

friedmud commented:
Ugh... I'm still against this whole `FEProblem`->`FEProblemBase` thing... again... there is a lot of user code with `FEProblem` in it...

friedmud commented:
Ok - I see now why it's necessary.  I don't see much of a way around it.

Is there any hope that Git is going to do the right thing here to preserve the authorship of all of these thousands of lines of code?

Is there anything we can do to help it out (like add a commit where `FEProblem.C` is renamed to `FEProblemBase.C`)?

fdkong commented:
> Ugh... I'm still against this whole FEProblem->FEProblemBase thing... again... there is a lot of user code with FEProblem in it...

Could you be more specific? I did not quite get what you said here.  It is fine, if users want to keep using  `FEProblem` (even though it is better to use `FEProblemBase`, in my opinion).

fdkong commented:
> Ok - I see now why it's necessary. I don't see much of a way around it.
> 
> Is there any hope that Git is going to do the right thing here to preserve the authorship of all of these thousands of lines of code?
> 
> Is there anything we can do to help it out (like add a commit where FEProblem.C is renamed to FEProblemBase.C)?

I have a commit `Rename FEProblem.{C,h} to FEProblemBase.{C, h} 			03f48cd`.  Is it not something you are looking for?

fdkong committed:
Address the comments and changed SLEPC safeguard

in NonlinearEigenSystem.{C,h} and EigenProblem.{C,h}

Refs #7398

friedmud commented:
@fdkong It looks like the authorship and paper-trail _are_ intact.  See here: https://github.com/fdkong/moose/blame/nonlinear_eigen_redesign_system_4/framework/src/base/FEProblemBase.C

(then click on "Blame" in the upper right... and let it load)

That was my last objection... so I think this is probably ready to go in.  Any further work can be done as future PRs.

Thanks for hanging in there and putting so much effort into this!

YaqiWang commented:
I just sent in a PR could interfering with this. Which one should be getting merged first?

permcody commented:
@YaqiWang - I'd prefer to merge this one first. It's huge, it's been reviewed. It's getting difficult to see small incremental changes at this point. Let's have you rebase your PR after this one is in.

permcody commented:
![image](https://cloud.githubusercontent.com/assets/3036960/20541870/29b5cb50-b0bc-11e6-97ef-2d2e4b3294c9.png)

andrsd commented:
Yeah, it is just 10K lines changed in 333 (half-the-devil) files...

YaqiWang commented:
Congratulations! This will also change my habit of browsing through moose.

fdkong commented:
Thanks a lot for @andrsd's time on reviewing and merging this tiny change.

Thanks a lot for comments from @friedmud, @permcody and @YaqiWang.

Thanks a lot for @jwpeterson's help on splitting commits.
